src/pymor/algorithms/rand_la.py

```diff
@@ -0,0 +1,362 @@
+        Q = A.apply(Q)
+        gram_schmidt(Q, range_product, atol=0, rtol=0, copy=False)
+
+    if return_rand:
+        return Q, R
+    else:
+        return Q
+
+
+@defaults('p', 'q', 'modes')
+def random_generalized_svd(A, range_product=None, source_product=None, modes=3, p=20, q=2):
+    """Randomized SVD of a |VectorArray| or an |Operator|. 
+
+    Saibaba AK, Hart J, van BloemenWaanders B. Randomized algorithms for
+    generalized singular value decomposition with application to sensitivity analysis. Numer Linear Algebra Appl.
+    2021;28:e2364. 
+    https://doi.org/10.1002/nla.2364
+
+    Viewing the 'A' as a 'A.dim' x 'len(A)' matrix, the return value
+    of this method is the randomized singular value decomposition of 'A', where the
+    inner product on R^('dim(A)') is given by 'range_product' and the inner product on R^('len(A)')
+    is given by 'source_product'. 
+        
+        A = U*\Sigma*V*source_product
+    
+    Parameters
+    ----------
+    A : 
+        The |VectorArray| or |Operator| for which the randomized SVD is to be computed. 
+    range_product : 
+        Range product |Operator| w.r.t which the randomized SVD is computed
+    source_product :
+        Source product |Operator| w.r.t which the randomized SVD is computed
+    modes : 
+        The first 'modes' approximated singular values and vectors are returned. 
+    p :
+        If not '0', adds 'Oversampling' colums to the randomly sampled matrix. 
+    q : 
+        If not '0', performs 'q' so called power iterations to increase the relative weight
+        of the first singular values.
+
+    Returns
+    -------
+    U 
+        |VectorArray| of approximated left singular vectors 
+    s 
+        One-dimensional |NumPy array| of the approximated singular values
+    V
+        |VectorArray| of the approximated right singular vectors
+    """
+    logger = getLogger('pymor.algorithms.rand_la')
+    
+    assert isinstance(A, VectorArray) or isinstance(A, Operator)
+    
+    if isinstance(A, VectorArray):
+        A = VectorArrayOperator(A)
+   
+    assert 0 <= modes <= max(A.source.dim, A.range.dim) and isinstance(modes, int)
+    assert p >= 0 and isinstance(p, int)
+    assert q >= 0 and isinstance(q, int)
+    
+    if range_product is None:
+        range_product = IdentityOperator(A.range)
+    else:
+        assert isinstance(range_product, Operator)
+        
+    if source_product is None:
+        source_product = IdentityOperator(A.source)
+    else:
+        assert isinstance(source_product, Operator)
+        
+    if A.source.dim == 0 or A.range.dim == 0:
+        return A.source.empty(), np.array([]), A.range.empty()
+    
+    Q = rrf(A, source_product=source_product, range_product=range_product, q=q, l=modes+p)
+    B = A.apply_adjoint(range_product.apply(Q))
+    Q_B, R_B = gram_schmidt(source_product.apply_inverse(B), product=source_product, return_R=True)
+    U_b, s, Vh_b = sp.linalg.svd(R_B.T)
```

Comment:
I think you could only compute the reduced matrices here to be more efficient.
```suggestion
    U_b, s, Vh_b = sp.linalg.svd(R_B.T, full_matrices=False)
```

Reply:
Good idea, i will change it.